upup/pkg/fi/cloudup/openstack/verifier.go

```diff
@@ -80,11 +81,46 @@ func NewOpenstackVerifier(opt *OpenStackVerifierOptions) (bootstrap.Verifier, er
 	}, nil
 }
 
+func validateJWTToken(token string) (jwt.MapClaims, error) {
+	certBytes, err := os.ReadFile("/etc/kubernetes/kops-controller/pki/bootstrap-ca.crt")
+	if err != nil {
+		return nil, fmt.Errorf("reading certificate: %v", err)
+	}
+
+	key, err := jwt.ParseRSAPublicKeyFromPEM(certBytes)
+	if err != nil {
+		return nil, fmt.Errorf("validate: parse key: %w", err)
+	}
+	tok, err := jwt.Parse(token, func(jwtToken *jwt.Token) (interface{}, error) {
+		if _, ok := jwtToken.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("unexpected method: %s", jwtToken.Header["alg"])
+		}
+		return key, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("validate: %w", err)
+	}
+	claims, ok := tok.Claims.(jwt.MapClaims)
+	if !ok || !tok.Valid {
+		return nil, fmt.Errorf("validate: invalid")
+	}
+	return claims, nil
+}
+
 func (o openstackVerifier) VerifyToken(ctx context.Context, token string, body []byte, useInstanceIDForNodeName bool) (*bootstrap.VerifyResult, error) {
 	if !strings.HasPrefix(token, OpenstackAuthenticationTokenPrefix) {
 		return nil, fmt.Errorf("incorrect authorization type")
 	}
-	serverID := strings.TrimPrefix(token, OpenstackAuthenticationTokenPrefix)
+	token = strings.TrimPrefix(token, OpenstackAuthenticationTokenPrefix)
+	tmp := strings.Split(token, ":")
+	if len(tmp) != 2 {
+		return nil, fmt.Errorf("invalid token")
+	}
+	serverID := tmp[0]
```

Comment:
Would there be value in putting the serverID into the claims? Or is it not available at the time we have to construct the JWT?

Reply:
its not available when constructing the JWT, the only thing that is available is instancegroup name.

Reply:
Okay, thanks for the info. I'm doubting the additional check is worth the effort.

Reply:
Well actually all other cloudproviders than openstack do use servergroups which makes it impossible to generate user_data per server. This means that we do not get big benefit of doing that for OpenStack only, because I am pretty sure that for instance hetzner (+scaleway?) is going to use this in instancegroup level. Also we have chicken - egg problem here. We cannot have serverID before the server is provisioned, we cannot either provision server before we have user_data. If we use server name instead, that is something people can modify (at least in openstack). So basically server name is not way to authorise.